deploy/runtools/run_farm_deploy_managers.py

```diff
@@ -50,6 +60,113 @@ def get_nbd_for_imagename(self, imagename: str) -> str:
 
         return self.allocated_dict[imagename]
 
+class URIContainer:
+    """ A class which contains the details for downloading a single URI. InstanceDeployManager
+    contains a list of this class. """
+
+    """a property name on RuntimeHWConfig"""
+    hwcfg_prop: str
+    """ the final filename inside sim_slot_x, this is a filename, not a path"""
+    destination_name: str
+
+    def __init__(self, hwcfg_prop: str, destination_name: str):
+        self.hwcfg_prop = hwcfg_prop
+        self.destination_name = destination_name
+
+    # this filename will be used when pre-downloading
+    @classmethod
+    def hashed_name(cls, uri) -> str:
+        m = hashlib.sha256()
+        m.update(bytes(uri, 'utf-8'))
+        return m.hexdigest()
+
+    def __resolve_vanilla_path(self, hwcfg) -> Optional[str]:
+        """ Allows fallback to a vanilla path. Relative paths are resolved realtive to firesim/deploy/. 
+        This will convert a vanilla path to a URI, or return None."""
+        uri: Optional[str] = getattr(hwcfg, self.hwcfg_prop)
+
+        # do nothing if there isn't a URI
+        if uri is None:
+            return None
+        
+        # if already a URI, exit early returning unmodified string
+        is_uri = re.match(_RFC_3986_PATTERN, uri)
+        if is_uri:
+            return uri
+
+        # expanduser() is required to get ~ home directory expansion working
+        # relative paths are relative to firesim/deploy
+        expanded = Path(expanduser(uri))
+
+        try:
+            # strict=True will throw if the file doesn't exist
+            resolved = expanded.resolve(strict=True)
+        except FileNotFoundError as e:
+            raise Exception(f"{self.hwcfg_prop} file fallback at path '{uri}' or '{expanded}' was not found")
+
+        return f"file://{resolved}"
+
+    def __choose_path(self, local_dir: str, hwcfg) -> Optional[Tuple[str, str]]:
+        """ Return a deterministic path, given a parent folder and a RuntimeHWConfig object. The URI
+        as generated from hwcfg is also returned. """
+        uri: Optional[str] = self.__resolve_vanilla_path(hwcfg)
+
+        # do nothing if there isn't a URI
+        if uri is None:
+            return None
+
+        # choose a repeatable, path based on the hash of the URI
+        destination = pjoin(local_dir, self.hashed_name(uri))
+        
+        return (uri, destination)
+
+    # download to a hashed destination name
+    # if the file exists this will not overwrite. Use this inside a 
```

Comment:
Is this comment duplicated? Its basically the same as the one inside the function